### PR TITLE
Sample without replacement & Finalise train/test split

### PR DESCRIPTION
**Pull request by johannes-kk**
- Add sampling without replacement
- Finalise train/test split implementation (sampling as shuffling)
- Remove some deprecated pointer usage
- Tidy formatting

### CODE REVIEW

**Review by hgupta18 — APPROVED** [review]
I don't fully understand the sampling function, but the updated train_test_split looks good to me.

**Comment by johannes-kk** [comment]
> I don't fully understand the sampling function, but the updated train_test_split looks good to me.

The updated `sample` function has two parts. If `replace == true` it uses the original code whereby it repeatedly samples the vector of row indices uniformly until the bootstrapped sample has `nrow` observations. If `replace == false` it instead shuffles the vector of original row indices, and pulls the first `nrow` from that, with at most the same number of observations as there are in the original dataframe.

`train_test_split` then just uses `sample` to bootstrap a sample with the same number of rows without replacement, meaning it effectively just shuffles the passed dataframe.

File: src/datasets.cpp [hunk 1]
```diff
@@ -520,23 +512,43 @@ DataFrame DataFrame::sample(int nrow, int seed) const{
...
+            // get random row index with replacement
+            int rand_row = distr(eng);
+            // get pointer to that row in original dataframe and store in bootstrap
+            new_frame.addRow(this->row(rand_row));
+        }
+    }else{
+        // pre-allocate vector of row indices
+        std::vector<int> indices(this->length());
+        // fill vector, equivalent to np.arange(0, this->length()-1)
+        std::generate(indices.begin(), indices.end(), [n = 0] () mutable {return n++;});
```
Comment (wfseaton) [inline src/datasets.cpp:541]: Per [this SO](https://stackoverflow.com/questions/50798713/c-generating-random-vector), you may be able to use distr() to directly generate random numbers when filling indices. This might save time/code, but current implementation works.

Comment (johannes-kk) [reply src/datasets.cpp:541]: If you mean the `distr` in the preceding `if` statement, I avoided using that as it just samples from the same uniform distribution each time, meaning it's effectively using replacement. In the highlighted code it shuffles the indices so it can pop them one at a time without replacing.

That said, I might full well be misunderstanding you. In either case, I've no deep affection for the code, so if you have a better suggestion (which seems probable) I'm more than open to improvements! 😀
